src/pythie_serving/lightgbm_wrapper.py

```diff
@@ -1,61 +1,98 @@
+                    "pandas_categorical": pandas_categorical,
+                }
 
     def Predict(self, request: predict_pb2.PredictRequest, context: grpc.RpcContext):
         model_name = request.model_spec.name
         if model_name not in self.model_map:
-            raise PythieServingException(f'Unknown model: {model_name}. This pythie-serving instance can only '
-                                         f'serve one of the following: {",".join(self.model_map.keys())}')
+            raise PythieServingException(
+                f"Unknown model: {model_name}. This pythie-serving instance can only "
+                f'serve one of the following: {",".join(self.model_map.keys())}'
+            )
 
         model_dict = self.model_map[model_name]
-
-        features_names = model_dict['feature_names']
-        samples = None
+        model = model_dict["model"]
+        pandas_categorical = model_dict["pandas_categorical"]
+        features_names = model_dict["feature_names"]
+        pd_categorical_features: List[str] = []
+        samples: Optional[List[List[Any]]] = None
         for feature_name in features_names:
             if feature_name not in request.inputs:
-                raise PythieServingException(f'{feature_name} not set in the predict request')
+                raise PythieServingException(
+                    f"{feature_name} not set in the predict request"
+                )
 
             nd_array = make_ndarray_from_tensor(request.inputs[feature_name])
+            # get features categories from model if feature dtype is object ("string")
+            is_pd_categorical = False
+            feature_categories = []
+            if nd_array.dtype == object:
+                if pandas_categorical is None:
+                    raise PythieServingException(f"{feature_name} feature has type 'object' but "
+                                                 f"there is no saved pandas categories from model")
+                is_pd_categorical = True
+                # lgbm save categories in the same order categorical features appear in model.feature_name()
+                feature_categories = pandas_categorical[
+                    len(pd_categorical_features)
+                ]
+                pd_categorical_features.append(feature_name)
+
             if len(nd_array.shape) != 2 or nd_array.shape[1] != 1:
-                raise PythieServingException('All input vectors should be 1D tensor')
+                raise PythieServingException("All input vectors should be 1D tensor")
 
-            if samples is None:
-                samples = [[] for _ in range(nd_array.shape[0])]
+            samples = samples or [[] for _ in range(nd_array.shape[0])]
 
             for sample_index, value in enumerate(nd_array):
-                samples[sample_index].append(value[0])
+                sample_feature_value = (
+                    feature_categories.index(value[0].decode("utf-8"))
```

Comment:
nothing guarantee that this was the encoding used, it would be safer to work with bytes everywhere - why do u need to decode ?  

Also the .index() function is costly as it is o(n) - can you use a dict instead to have o(1) lookup here ?

Reply:
We need to decode because `value[0]` is a bytestring and we need to convert it into a string as pandas categories are string values. In fact the client will be force to encode its string data in "utf-8".